res/emptty.1
.TH EMPTTY 1 "March 2026" "emptty 0.16.1" emptty

.SH NAME
emptty \- Dead simple CLI Display Manager on TTY

.SH SYNOPSIS
.B emptty [-v] [--version] [-d] [--daemon] [-c PATH] [--config PATH] [-i] [--ignore-config] [-t TTY] [--tty TTY] [-u defaultUser] [--default-user defaultUser] [-a [session]] [--autologin [session]]

.SH DESCRIPTION
.B emptty
Simple CLI Display Manager, that allows one to select DE/WM after login, use predefined config or allows autologin, if selected user is part of
.I nopasswdlogin
group.

.SH OPTIONS
.IP "\-v, \-\-version"
Display the version of the program.

.IP "\-d, \-\-daemon"
Starts emptty as daemon, that does not require agetty.

.IP "\-c, \-\-config PATH"
Loads configuration from specified path.

.IP "\-C, \-\-print\-config"
Only prints loaded configuration and exits.

.IP "\-i, \-\-ignore-config"
Skips loading of configuration from file, loads only argument configuration.

.IP "\-t, \-\-tty TTY"
Overrides loaded configuration by setting defined TTY. May be specified as a number (e.g. "7") or a TTY name (e.g. "tty7").

.IP "\-u, \-\-default-user defaultUser"
Overrides loaded configuration by setting defined defaultUser.

.IP "\-a, \-\-autologin [session]"
Overrides loaded configuration by enabling autologin. If session is defined, it overrides autologin session.

.SH CONFIG
/etc/emptty/conf

.IP TTY_NUMBER
TTY, where emptty will start.
.IP SWITCH_TTY
Enables switching to defined TTY number. Default is true.
.IP PRINT_ISSUE
Enables printing of /etc/issue in daemon mode.
.IP PRINT_MOTD
Enables printing of default motd, static motd or dynamic motd.
.IP DEFAULT_ENV
Defines default environment used for starting undefined sessions (e.g. from `emptty` file). Possible values are "xorg" and "wayland". Default of default is xorg.
.IP DEFAULT_USER
Preselected user, if AUTOLOGIN is enabled, this user is logged in.
.IP DEFAULT_SESSION
Preselected desktop session, if user does not use `emptty` file. Has lower priority than
.I AUTOLOGIN_SESSION
.IP DEFAULT_SESSION_ENV
Optional environment of preselected desktop session, if user does not use `emptty` file. Possible values are "xorg" and "wayland".
.IP AUTOLOGIN
Enables Autologin, if DEFAULT_USER is defined. Possible values are "true" or "false". Default value is false.

.B NOTE:
to enable autologin DEFAULT_USER must be in group
.I nopasswdlogin
, otherwise user will NOT be authorized.
.IP AUTOLOGIN_SESSION
The default session used, if Autologin is enabled. If session is not found in list of session, it proceeds to manual selection.
.IP AUTOLOGIN_SESSION_ENV
Optional environment of autologin desktop session. Possible values are "xorg" and "wayland".
.IP AUTOLOGIN_MAX_RETRY
If session does not start correctly in specified period, the number of retries in short period is kept to eventually stop the infinite loop of restarts. -1 is for infinite retries, 0 is for no retry. Default value is 2.
.IP AUTOLOGIN_RETRY_PERIOD
Time period in seconds, that is used for checking session login retries. Default value is 2.
.IP LANG
defines locale for all users. Default value is "en_US.UTF-8"
.IP DBUS_LAUNCH
Starts "dbus-launch" before desktop command. After end of session "dbus-daemon" is interrupted. Default value is true. If
.I user config
is handled as script (does not contain
.I Exec
option), this config is overridden to false.
.IP ALWAYS_DBUS_LAUNCH
Starts "dbus-launch" before desktop command in any case,
.I DBUS_LAUNCH
value is ignored. It also starts even if
.I
XINITRC_LAUNCH
is set to `true`. After end of session "dbus-daemon" is interrupted. Default value is false.
.IP XINITRC_LAUNCH
Starts Xorg desktop with calling
.I ~/.xinitrc
script with session exec as argument, if is true, file exists and selected WM/DE is Xorg session, it overrides DBUS_LAUNCH. If
.I user config
is handled as script (does not contain
.I Exec
option), this config is overridden to false.
.IP VERTICAL_SELECTION
Prints available WM/DE each on new line instead of printing on single line.
.IP INDENT_SELECTION
Horizontal offset for vertical selection, useful for centering the menu on screen.
.IP LOGGING
Defines how logging is handled. Possible values are "rotate", "appending" or "disabled". Default value is "rotate".
.IP LOGGING_FILE
Overrides path of log file. Default value is
.I /var/log/emptty/[TTY_NUMBER].log

.B NOTE:
It expects existence of directories to defined logging file.

.IP XORG_CMD
Overrides path to the X server executable, allowing use of different X server implementations.
.IP XORG_ARGS
Arguments passed to Xorg server.
.IP DYNAMIC_MOTD
Allows using dynamic motd script to generate custom MOTD. Possible values are "true" or "false". Default value is false.

.B NOTE:
Be sure, that dynamic motd has correct content and permissions (e.g. 744), the script is started as default user; in daemon mode it means
.I root

.IP DYNAMIC_MOTD_PATH
Allows overriding default path to dynamic motd. Default value is "/etc/emptty/motd-gen.sh".

.IP MOTD_PATH
Allows overriding default path to static motd. Default value is "/etc/emptty/motd".

.IP FG_COLOR
Foreground color, available only in daemon mode. List of colors is listed below.

.IP BG_COLOR
Background color, available only in daemon mode. List of colors is listed below.

.IP DISPLAY_START_SCRIPT
Script started before Display (Xorg/Wayland) starts.

.B NOTE:
The script is started as default user; in daemon mode it means
.I root

.IP DISPLAY_STOP_SCRIPT
Script started after Display (Xorg/Wayland) stops.

.B NOTE:
The script is started as default user; in daemon mode it means
.I root

.IP ENABLE_NUMLOCK
Enables numlock in daemon mode. Possible values are "true" or "false". Default value is false.

.IP SESSION_ERROR_LOGGING
Defines how logging of session errors is handled. Possible values are "rotate", "appending" or "disabled". Default value is "disabled".
.IP SESSION_ERROR_LOGGING_FILE
Overrides path of session errors log file. Default value is
.I /var/log/emptty/session-errors.[TTY_NUMBER].log

.B NOTE:
It expects existence of directories to defined logging file.
.IP NO_XDG_FALLBACK
Disallows setting of fallback values for all XDG environmental variables and leaves it on Login Controls. Possible values are "true" or "false". Default value is false.

.B NOTE:
Be aware, that setting to "true" could lead to unexpected behaviour.

.IP DEFAULT_XAUTHORITY
If set true, it will not use .emptty-xauth file, but the standard ~/.Xauthority file. This allows handling xauth issues. Possible values are "true" or "false". Default value is false.

.IP ROOTLESS_XORG
If set true, Xorg will be started as rootless, if system allows and emptty is running in daemon mode. Possible values are "true" or "false". Default value is false.

.IP IDENTIFY_ENVS
If set true, environmental groups are printed to differ Xorg/Wayland/Custom/UserCustom desktops. Possible values are "true" or "false". Default value is false.

.IP HIDE_ENTER_LOGIN
If set true, "hostname login:" is not displayed. Possible values are "true" or "false". Default value is false.

.IP HIDE_ENTER_PASSWORD
If set true, "Password:" is not displayed. Possible values are "true" or "false". Default value is false.

.IP XORG_SESSIONS_PATH
Path to directory, where Xorg sessions' desktop files are stored. Default value is "/usr/share/xsessions/".

.IP WAYLAND_SESSIONS_PATH
Path to directory, where Wayland sessions' desktop files are stored. Default value is "/usr/share/wayland-sessions/".

.IP SELECT_LAST_USER
Enables funtionality of saving last successfully logged in user for next login. Possible values are "false", "per-tty" or "global". Default value is false.

.IP AUTO_SELECTION
If set to "true" and only one desktop is available, it automatically select that desktop. Possible values are "true" or "false". Default value is false.

.IP ALLOW_COMMANDS
If set to "true" and no default user is selected, it allows to enter commands into login input. Possible values are "true" or "false", Default value is true.

.IP CMD_POWEROFF
Command to be used to perform poweroff. Default value is "poweroff".

.IP CMD_REBOOT
Command to be used to perform reboot. Default value is "reboot".

.IP CMD_SUSPEND
Command to be used to perform suspend. Default value is blank, but it tries to use "systemctl suspend", "loginctl suspend" or "zzz".

.IP WAIT_EXIT_TIMEOUT
Timeout in seconds before emptty automatically exits. If value is 0 or lower, there is no timeout. Default value is -1.

.IP WAIT_CARD_DEVICE_COUNT
Number of card devices to wait for before starting. If value is 0 or lower, there is no wait. Default value is 0.

.SH COMMANDS
If commands are allowed and default user is not defined, there could be used commands in login input. All of these commands need to start with colon ":". Escape characters are ignored to prevent issues with muscle memory from VI.
 - :help, :? - prints available commands
 - :poweroff, :shutdown - processess poweroff command
 - :reboot - processes reboot command
 - :suspend, :zzz - processes suspend command

.SH DYNAMIC MOTD
Optional file stored by default as /etc/emptty/motd-gen.sh. Could be overridden.

If
.I DYNAMIC_MOTD
is set to true, this file exists and is executable for its owner, the result is printed as your own MOTD. Be very careful with this script!

.SH CUSTOM MOTD
Optional file stored by default as /etc/emptty/motd. Could be overridden.

Custom file, that prints your own MOTD. Reading this file supports colors (e.g.
.I \\\x1b[31m
or
.I \\\033[32m
)

.SH USER CONFIG
Optional file stored as ${HOME}/.config/emptty or ${HOME}/.emptty

Configuration file stored as ${HOME}/.config/emptty has higher priority on loading.
.IP Name
Optional name to be used as Session Name.
.IP Exec
Defines command to start Desktop Environment/Window Manager. This value does not need to be defined, if user config is presented as shell script (with shebang at the start and execution permissions).
.IP Environment
Selects, which environment should be defined for following command. Possible values are "xorg" and "wayland", "xorg" is default.
.IP Lang
Defines locale for logged user, has higher priority than LANG from global configuration
.IP Selection
Requires selection of desktop, basically turns
.I emptty
file into
.I .xinitrc
for Xorg and Wayland. In this case
.I Exec
is skipped. Possible values are "false" for never using selection, "true" for always showing selection or "auto" for showing selection or first option autoselect, if there is no other desktop. Default value is false.
.IP LoginShell
Defines custom shell to be used to start the session. This allows starting the session with non-interactive shell e.g. "/bin/bash --login"
.IP DesktopNames
Value passed into
.I XDG_CURRENT_DESKTOP
variable.

.SH USER EXIT SCRIPT
Optional script file stored as ${HOME}/.config/emptty-exit, that is handled as shell script and is started, when session is going to end. Script is started even if emptty is being terminated.
The default timeout to finish script is 3 seconds, but it is configurable from the script itself.
.IP Timeout
Optional custom timeout for script to finish its run, number represents seconds. Default is 3.

.SH CUSTOM SESSIONS
Optional folders for custom sessions, that could be available system-wide (in case of /etc/emptty/custom-sessions/) or user-specific (in case of ${HOME}/.config/emptty-custom-sessions/), but do not have .desktop file stored on standard paths for Xorg or Wayland sessions. Expected suffix of each file is ".desktop".

.IP Name
Defines name of Desktop Environment/Window Manager.
.IP Exec
Defines command to start Desktop Environment/Window Manager. It could contain multiple arguments same as in *.desktop files.
.IP Environment
Selects, which environment should be defined for following command. Possible values are "xorg" and "wayland", "xorg" is default.
.IP DesktopNames
Value passed into
.I XDG_CURRENT_DESKTOP
variable.
.IP NoDisplay/Hidden
Boolean value, that controls visibility of desktop session.

.SH LAST SESSION
The last user selection of session is stored into ~/.cache/emptty/last-session

.SH LOGGING
As it is mentioned in configuration, there are three options to handle logging of emptty. The logs contains not just logs from emptty, but also from Xorg (if used) and user's WM/DE.
Described log location could differ according configuration
.I LOGGING_FILE
, that is stored in
.I /etc/emptty/conf

.IP default
This option provides simple solution, when current instance of emptty logs into
.I /var/log/emptty/[TTY_NUMBER].log
and the previous version is stored as
.I /var/log/emptty/[TTY_NUMBER].log.old

.B NOTE:
Current instance always move previous log into old file, if emptty crashes and is started again, previous log is in
.I /var/log/emptty/[TTY_NUMBER].log.old

.IP appending
This option provides functionality that logs everything into
.I /var/log/emptty/[TTY_NUMBER].log
and does not handle log rotation by itself. It leaves the option for user to handle it themselves (e.g. with logrotate).

.B NOTE:
Appending without roration could cause large log file, be sure that log file is rotated.

.IP disabled
This option points all log into
.I /dev/null
, so no log is available.

.B NOTE:
If any issue starts to appear and you want to report it, ensure you do not use this option.

.SH COLORS
Please, be aware that
.I LIGHT_
colors could be unavailable as background color.

BLACK, RED, GREEN, YELLOW, BLUE, PURPLE, CYAN, WHITE

LIGHT_BLACK, LIGHT_RED, LIGHT_GREEN, LIGHT_YELLOW, LIGHT_BLUE, LIGHT_PURPLE, LIGHT_CYAN, LIGHT_WHITE


src/emptty.go
package src

import (
	"errors"
	"fmt"
	"os"
	"os/signal"
	"runtime"
	"strings"
	"syscall"
)

const (
	version = "0.16.1"

	builtinCmdHelpString = `
Available commands:
  :help, :?			print this help
  :poweroff, :shutdown		process poweroff command
  :reboot			process reboot command
  :suspend, :zzz		process suspend command
`
	terminalHelpString = `Usage: emptty [options]
Options:
  -h, --help			print this help
  -v, --version			print version
  -d, --daemon			start in daemon mode
  -c, --config PATH		load configuration from specified path
  -C, --print-config		prints currently loaded configuration
  -i, --ignore-config		skips loading of configuration from file, loads only argument configuration
  -t, --tty NUMBER		overrides configured TTY number
  -u, --default-user USER_NAME	overrides configured Default User
  -a, --autologin [SESSION]	overrides configured autologin to true and if next argument is defined, it defines also Autologin Session
`
)

var buildVersion string
var errPrintCommandHelp = errors.New("just print help")

type sessionHandle struct {
	session     *commonSession
	auth        authHandle
	interrupted bool
}

func init() {
	runtime.LockOSThread()
}

// Main handles the functionality of whole application.
func Main() {
	TEST_MODE = false

	processCoreArgs(os.Args)

	conf := loadConfig(loadConfigPath(os.Args))
	processArgs(os.Args, conf)

	fTTY := startDaemon(conf)

	initLogger(conf)
	waitForCardDevice(conf)
	printMotd(conf)

	if command := login(conf, initSessionHandle()); command != "" {
		processCommand(command, conf, nil, false)
	}

	stopDaemon(conf, fTTY)
}

// Initialize session handle with common interrupt handler
func initSessionHandle() *sessionHandle {
	h := &sessionHandle{}

	c := make(chan os.Signal, 10)
	signal.Notify(c, os.Interrupt, syscall.SIGHUP, syscall.SIGINT, syscall.SIGQUIT, syscall.SIGTERM)
	go handleInterrupt(c, h)

	return h
}

// Catch interrupt signal chan and interrupts Cmd.
func handleInterrupt(c chan os.Signal, h *sessionHandle) {
	<-c
	logPrint("Caught interrupt signal")
	setTerminalEcho(os.Stdout.Fd(), true)
	h.interrupted = true

	if h.session != nil && h.session.cmd != nil {
		h.session.interrupted = true
		if err := h.session.cmd.Process.Signal(os.Interrupt); err != nil {
			logPrint("Application not responding to signal")
		} else {
			// Only attempt clean shutdown if cmd is still
			// responsive
			h.session.cmd.Wait()
		}
	}
	// Always attempt to close active auth instances before exit
	if h.auth != nil {
		h.auth.closeAuth()
	}
	os.Exit(1)
}

// Process core arguments for help and version, because they don't require any further application run
func processCoreArgs(args []string) {
	if contains(args, "-h", "--help") {
		printHelp()
		os.Exit(0)
	}
	if contains(args, "-v", "--version") {
		fmt.Printf("emptty %s\nhttps://github.com/tvrzna/emptty\n\nReleased under the MIT License.\n", getVersion())
		os.Exit(0)
	}
}

// Loads config from path according to arguments
func loadConfigPath(args []string) (configPath string) {
	configPath = pathConfigFile

	for i, arg := range args {
		switch arg {
		case "-c", "--config":
			nextArg(args, i, func(val string) {
				if fileExists(val) {
					configPath = val
				}
			})
			return configPath
		case "-i", "--ignore-config":
			return ""
		}
	}

	return configPath
}

// Process arguments with affection on configuration
func processArgs(args []string, conf *config) {
	printConfig := false

	for i, arg := range args {
		switch arg {
		case "-t", "--tty":
			nextArg(args, i, func(val string) {
				tty := parseTTY(val, "0")
				if tty > 0 {
					conf.Tty = tty
				} else {
					ttynum := strings.SplitAfterN(val, "tty", 2)
					if len(ttynum) == 2 {
						tty := parseTTY(ttynum[1], "0")
						if tty > 0 {
							conf.Tty = tty
						}
					}
				}
			})
		case "-u", "--default-user":
			nextArg(args, i, func(val string) {
				conf.DefaultUser = val
			})
		case "-d", "--daemon":
			conf.DaemonMode = true
		case "-a", "--autologin":
			conf.Autologin = true
			nextArg(args, i, func(val string) {
				conf.AutologinSession = val
			})
		case "-C", "--print-config":
			printConfig = true
		}
	}

	if printConfig {
		conf.printConfig()
		os.Exit(0)
	}
}

// Gets next argument, if available
func nextArg(args []string, i int, callback func(value string)) {
	if callback != nil && len(args) > i+1 {
		val := sanitizeValue(args[i+1], "")
		if !strings.HasPrefix(val, "-") {
			callback(args[i+1])
		}
	}
}

// Prints help
func printHelp() {
	fmt.Print(terminalHelpString)
}

// Gets current version
func getVersion() string {
	tags := strings.Builder{}
	for _, tag := range []string{tagPam, tagUtmp} {
		if tags.Len() > 0 {
			tags.WriteString(", ")
		}
		tags.WriteString(tag)
	}
	if buildVersion != "" {
		if tags.Len() == 0 {
			return buildVersion[1:]
		}
		return buildVersion[1:] + " (" + tags.String() + ")"
	}
	return version
}

func formatCommand(command string) string {
	return strings.ReplaceAll(command, "\x1b", "")[1:]
}

func shouldProcessCommand(input string, conf *config) bool {
	return conf.AllowCommands && strings.HasPrefix(strings.ReplaceAll(input, "\x1b", ""), ":")
}

// Process commands input in login buffer
func processCommand(command string, c *config, auth authHandle, continuable bool) error {
	switch command {
	case "help", "?":
		fmt.Print(builtinCmdHelpString)
		if continuable {
			return errPrintCommandHelp
		}
		waitForReturnToExit(0)
	case "poweroff", "shutdown":
		if continuable && auth != nil {
			auth.closeAuth()
		}
		if err := processCommandAsCmd(c.CmdPoweroff); err == nil {
			waitForReturnToExit(0)
		} else {
			handleErr(err)
		}
	case "reboot":
		if continuable && auth != nil {
			auth.closeAuth()
		}
		if err := processCommandAsCmd(c.CmdReboot); err == nil {
			waitForReturnToExit(0)
		} else {
			handleErr(err)
		}
	case "suspend", "zzz":
		if continuable && auth != nil {
			auth.closeAuth()
		}
		variants := []string{
			"zzz",
			"systemctl suspend",
			"loginctl suspend",
		}
		if c.CmdSuspend != "" {
			variants = append([]string{c.CmdSuspend}, variants...)
		}

		var err error
		for _, v := range variants {
			if err = processCommandAsCmd(v); err != nil {
				continue
			} else {
				break
			}
		}

		if err == nil {
			waitForReturnToExit(0)
		} else {
			handleErr(err)
		}
	default:
		err := fmt.Errorf("Unknown command '%s'", command)
		if continuable {
			return err
		}
		handleErr(err)
	}
	return nil
}
